Jenkinsfile

```diff
@@ -0,0 +1,369 @@
+def notify(status){
```

Comment:
We won't be carrying the Jenkins config as it won't be exercised by our build system, but we would very much appreciate some help setting up a script that can automate building, preferably via CircleCI (but any other services that integrate with GitHub for free will do). I have done some of the work on it but it's not fully working:
```
  m68k-amigaos:
    docker:
      - image: amigadev/crosstools:m68k-amigaos
    working_directory: ~/repo
    steps:
      - checkout
      - run: export TARGET='m68k-amigaos'
      - run: export SYSROOT=/opt/$TARGET
      - run: export FLAGS='-m68080 -mhard-float -s -ffast-math -O3 -w -fomit-frame-pointer  -noixemul -DBIG_ENDIAN -DAMIGA -fpermissive -std=c++14 -fbbb=abcdef'
      - run: export DEFINES='-DUSE_SDL1=ON -DNONET=ON -DM68K_CPU=68040 -DM68K_FPU=hard -DM68K_COMMON="-s -ffast-math -O3 -fno-exceptions -w -fomit-frame-pointer  -noixemul -DBIG_ENDIAN -DAMIGA -fpermissive -std=c++14 -fbbb=abcdef"'
      - run: wget https://www.zlib.net/zlib-1.2.11.tar.gz
      - run: tar -xvf zlib-1.2.11.tar.gz
      - run: wget https://raw.githubusercontent.com/Kitware/CMake/master/Modules/FindZLIB.cmake -O CMake/FindZLIB.cmake
      - run: mkdir -p zlib-1.2.11/build
      - run: cd zlib-1.2.11/build && eval cmake .. -DCMAKE_INSTALL_PREFIX=${SYSROOT} ${DEFINES}
      - run: cd zlib-1.2.11/build && cmake --build . --config Release --target install -- -j$(nproc)
      - run: wget https://github.com/AmigaPorts/libSDL12/archive/master.tar.gz -O SDL-1.2.tar.gz
      - run: tar -xvf SDL-1.2.tar.gz
      - run: cd libSDL12-master && make PREFX=${SYSROOT} PREF=${SYSROOT} -j$(nproc)
      - run: cd libSDL12-master && cp -fvr libSDL.a ${SYSROOT}/lib
      - run: cd libSDL12-master && cp -fvr include/* ${SYSROOT}/include/
      - run: wget https://github.com/SDL-mirror/SDL_mixer/archive/SDL-1.2.tar.gz -O SDL_mixer-SDL-1.2.tar.gz
      - run: tar -xvf SDL_mixer-SDL-1.2.tar.gz
      - run: cd SDL_mixer-SDL-1.2 && ./autogen.sh
      - run: cd SDL_mixer-SDL-1.2 && SDL_LIBS='-lSDL -ldebug' SDL_CFLAGS="-I${SYSROOT}/include/SDL -noixemul" ./configure --disable-sdltest --disable-shared --enable-static --host=${TARGET} --prefix=${SYSROOT}
      - run: cd SDL_mixer-SDL-1.2 && make -j$(nproc)
      - run: cd SDL_mixer-SDL-1.2 && make install
      - run: wget https://download.savannah.gnu.org/releases/freetype/freetype-2.10.1.tar.gz
      - run: tar -xvf freetype-2.10.1.tar.gz
      - run: mkdir -p freetype-2.10.1/build
      - run: cd freetype-2.10.1/build && eval cmake .. -DCMAKE_INSTALL_PREFIX=${SYSROOT} -DUNIX=1 ${DEFINES}
      - run: cd freetype-2.10.1/build && cmake --build . --config Release --target install -- -j$(nproc)
      - run: wget https://github.com/SDL-mirror/SDL_ttf/archive/SDL-1.2.tar.gz -O SDL_ttf-SDL-1.2.tar.gz
      - run: tar -xvf SDL_ttf-SDL-1.2.tar.gz
      - run: cd SDL_ttf-SDL-1.2 && ./autogen.sh
      - run: cd SDL_ttf-SDL-1.2 && CFLAGS="${FLAGS}" CXXFLAGS="${FLAGS}" SDL_LIBS='-lSDL -ldebug' SDL_CFLAGS="-I${SYSROOT}/include/SDL -noixemul" FT2_CFLAGS="-I${SYSROOT}/include/freetype2" FT2_LIBS="-lfreetype -lzlib" ./configure --disable-shared --enable-static --host=${TARGET} --prefix=${SYSROOT}
      - run: cd SDL_ttf-SDL-1.2 && make -j$(nproc)
      - run: cd SDL_ttf-SDL-1.2 && make install
      - run: wget https://raw.githubusercontent.com/Kitware/CMake/v3.10.0/Modules/SelectLibraryConfigurations.cmake -O CMake/SelectLibraryConfigurations.cmake
      - run: mkdir build
      - run: cd build && eval PKG_CONFIG_PATH=${SYSROOT}/lib/pkgconfig/:${SYSROOT}/share/pkgconfig/ cmake ${DEFINES} ..
      - run: cd build && cmake --build . --config Release -- -j$(nproc)
      - store_artifacts: {path: ./build/devilutionx, destination: devilutionx_m68k_amigaos}
```
(probably a helper script in .sh would be a good idea). This can also serve to generate daily builds that can be tested by users to see if the project is still working correctly.

Reply:
What's wrong? What fails?
I see that two different CPU types are defined 68080 + 68040, I bet that doesn't work out too well. :p

Reply:
Ad the very end when installing SDL_ttf:
```
make[1]: Entering directory '/SDL_ttf-SDL-1.2'
 /usr/bin/mkdir -p '/opt/m68k-amigaos/lib'
 /bin/bash ./libtool   --mode=install /usr/bin/install -c   libSDL_ttf.la '/opt/m68k-amigaos/lib'
libtool: install: /usr/bin/install -c .libs/libSDL_ttf.lai /opt/m68k-amigaos/lib/libSDL_ttf.la
libtool: install: /usr/bin/install -c .libs/libSDL_ttf.a /opt/m68k-amigaos/lib/libSDL_ttf.a
libtool: install: chmod 644 /opt/m68k-amigaos/lib/libSDL_ttf.a
libtool: install: /opt/m68k-amigaos/bin/m68k-amigaos-ranlib /opt/m68k-amigaos/lib/libSDL_ttf.a
/SDL_ttf-SDL-1.2/libtool: line 2278: /sys/libs: No such file or directory
----------------------------------------------------------------------
Libraries have been installed in:
   /opt/m68k-amigaos/lib

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the `-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the `LD_RUN_PATH' environment variable
     during linking
   - use the `-LLIBDIR' linker flag
   - have your system administrator add LIBDIR to `/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
 /usr/bin/mkdir -p '/opt/m68k-amigaos/include/SDL'
 /usr/bin/install -c -m 644 SDL_ttf.h '/opt/m68k-amigaos/include/SDL'
 /usr/bin/mkdir -p '/opt/m68k-amigaos/lib/pkgconfig'
 /usr/bin/install -c -m 644 SDL_ttf.pc '/opt/m68k-amigaos/lib/pkgconfig'
make[1]: Leaving directory '/SDL_ttf-SDL-1.2'
mkdir: cannot create directory 'build': File exists
```

Reply:
Here is a bash version of the script if you want to give it a spin:
`docker run -ti --rm -v $(pwd):/work amigadev/crosstools:m68k-amigaos bash`
```
#!/bin/bash

# exit when any command fails
set -e

#set compiler params
export TARGET='m68k-amigaos'
export SYSROOT=/opt/$TARGET
export FLAGS='-m68040 -mhard-float -s -ffast-math -O3 -w -fomit-frame-pointer  -noixemul -DBIG_ENDIAN -DAMIGA -fpermissive -std=c++14 -fbbb=abcdef'
export DEFINES='-DUSE_SDL1=ON -DNONET=ON -DM68K_CPU=68040 -DM68K_FPU=hard -DM68K_COMMON="-s -ffast-math -O3 -fno-exceptions -w -fomit-frame-pointer  -noixemul -DBIG_ENDIAN -DAMIGA -fpermissive -std=c++14 -fbbb=abcdef"'

# CMake helper scripts
wget https://raw.githubusercontent.com/Kitware/CMake/master/Modules/FindZLIB.cmake -O CMake/FindZLIB.cmake
wget https://raw.githubusercontent.com/Kitware/CMake/v3.10.0/Modules/SelectLibraryConfigurations.cmake -O CMake/SelectLibraryConfigurations.cmake

cd buildm

# ZLIB
wget https://www.zlib.net/zlib-1.2.11.tar.gz
tar -xvf zlib-1.2.11.tar.gz
mkdir -p zlib-1.2.11/build
cd zlib-1.2.11/build
eval cmake .. -DCMAKE_INSTALL_PREFIX=${SYSROOT} ${DEFINES}
cmake --build . --config Release --target install -- -j$(nproc)
cd ../..

# SDL1.2
wget https://github.com/AmigaPorts/libSDL12/archive/master.tar.gz -O SDL-1.2.tar.gz
tar -xvf SDL-1.2.tar.gz
cd libSDL12-master
make PREFX=${SYSROOT} PREF=${SYSROOT} -j$(nproc)
cp -fvr libSDL.a ${SYSROOT}/lib
cp -fvr include/* ${SYSROOT}/include/
cd ../..

# SDL_mixer
wget https://github.com/SDL-mirror/SDL_mixer/archive/SDL-1.2.tar.gz -O SDL_mixer-SDL-1.2.tar.gz
tar -xvf SDL_mixer-SDL-1.2.tar.gz
cd SDL_mixer-SDL-1.2
./autogen.sh
SDL_LIBS='-lSDL -ldebug' SDL_CFLAGS="-I${SYSROOT}/include/SDL -noixemul" ./configure --disable-sdltest --disable-shared --enable-static --host=${TARGET} --prefix=${SYSROOT}
make -j$(nproc)
make install
cd ../..

# FreeType
wget https://download.savannah.gnu.org/releases/freetype/freetype-2.10.1.tar.gz
tar -xvf freetype-2.10.1.tar.gz
mkdir -p freetype-2.10.1/build
cd freetype-2.10.1/build
eval cmake .. -DCMAKE_INSTALL_PREFIX=${SYSROOT} -DUNIX=1 ${DEFINES}
cmake --build . --config Release --target install -- -j$(nproc)
cd ../..

# SDL_ttf
wget https://github.com/SDL-mirror/SDL_ttf/archive/SDL-1.2.tar.gz -O SDL_ttf-SDL-1.2.tar.gz
tar -xvf SDL_ttf-SDL-1.2.tar.gz
cd SDL_ttf-SDL-1.2
./autogen.sh
CFLAGS="${FLAGS}" CXXFLAGS="${FLAGS}" SDL_LIBS='-lSDL -ldebug' SDL_CFLAGS="-I${SYSROOT}/include/SDL -noixemul" FT2_CFLAGS="-I${SYSROOT}/include/freetype2" FT2_LIBS="-lfreetype -lzlib" ./configure --disable-shared --enable-static --host=${TARGET} --prefix=${SYSROOT}
make -j$(nproc)
make install
cd ../..

# DevilutionX
mkdir build
cd build && eval PKG_CONFIG_PATH=${SYSROOT}/lib/pkgconfig/:${SYSROOT}/share/pkgconfig/ cmake ${DEFINES} ..
cd build && cmake --build . --config Release -- -j$(nproc)
```

Reply:
the error is not that SDL_ttf fails, it's that the build directory already exists.
```rm -rf build
mkdir build```
or if you want to keep the previously built files and just build the changes
```mkdir -p build```

Reply:
Derp, I was simply jumping up the wrong number of folders... glad I didn't `rm *` somewhere :D

Here is the first issue I was having last I attempted:
```
[ 80%] Building CXX object CMakeFiles/devilutionx.dir/SourceX/DiabloUI/diabloui.cpp.obj
/work/SourceX/dvlnet/packet.cpp: In member function 'const buffer_t& dvl::net::packet::message()':
/work/SourceX/dvlnet/packet.cpp:43:26: error: exception handling disabled, use -fexceptions to enable
   throw packet_exception();
                          ^
make[2]: *** [CMakeFiles/devilutionx.dir/build.make:245: CMakeFiles/devilutionx.dir/SourceX/dvlnet/packet.cpp.obj] Error 1
make[2]: *** Waiting for unfinished jobs....
/work/SourceX/dvlnet/frame_queue.cpp: In member function 'dvl::net::buffer_t dvl::net::frame_queue::read(dvl::net::framesize_t)':
/work/SourceX/dvlnet/frame_queue.cpp:18:31: error: exception handling disabled, use -fexceptions to enable
   throw frame_queue_exception();
                               ^
make[2]: *** [CMakeFiles/devilutionx.dir/build.make:271: CMakeFiles/devilutionx.dir/SourceX/dvlnet/frame_queue.cpp.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:77: CMakeFiles/devilutionx.dir/all] Error 2

```

Reply:
The error speaks for itself. -fexceptions must be added. I assume the default C-flag in my docker toolchain sets -fno-exceptions due to it making slower binaries on Amiga.

Reply:
I mentioned this in a comment above

Reply:
Removing all exception related code then get me to an error regarding _mbstate_t not being defined in stdlib.h, this is probably a bug in the m68k header files, but can besolved by adding `#include <iostream>`

At that point I get fully stuck at freetype wanting gzip during the link process:
```
[ 78%] Linking CXX executable devilutionx
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: /opt/m68k-amigaos/lib/libfreetype.a(ftgzip.c.o):(.text+0x54): undefined reference to `inflate'
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: /opt/m68k-amigaos/lib/libfreetype.a(ftgzip.c.o): in function `FT_Gzip_Uncompress':
�$��:32525: undefined reference to `inflate'
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: /opt/m68k-amigaos/lib/libfreetype.a(ftgzip.c.o):(.text+0x21e): undefined reference to `inflateEnd'
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: /opt/m68k-amigaos/lib/libfreetype.a(ftgzip.c.o): in function `FT_Stream_OpenGzip':
�$��:32525: undefined reference to `inflateEnd'
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: /opt/m68k-amigaos/lib/libfreetype.a(ftgzip.c.o): in function `FT_Gzip_Uncompress':
�$��:32525: undefined reference to `inflateEnd'
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: �$��:32525: undefined reference to `inflateEnd'
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: /opt/m68k-amigaos/lib/libfreetype.a(ftgzip.c.o): in function `FT_Stream_OpenGzip':
�$��:32525: undefined reference to `inflateReset'
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: /opt/m68k-amigaos/lib/libfreetype.a(ftgzip.c.o):(.text+0x1cc): undefined reference to `inflateReset'
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: /opt/m68k-amigaos/lib/libfreetype.a(ftgzip.c.o): in function `FT_Gzip_Uncompress':
�$��:32525: undefined reference to `inflateInit2_'
/opt/amiga/lib/gcc/m68k-amigaos/6.5.0b/../../../../m68k-amigaos/bin/ld: /opt/m68k-amigaos/lib/libfreetype.a(ftgzip.c.o): in function `FT_Stream_OpenGzip':
�$��:32525: undefined reference to `inflateInit2_'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/devilutionx.dir/build.make:517: devilutionx] Error 1
make[1]: *** [CMakeFiles/Makefile2:77: CMakeFiles/devilutionx.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

Reply:
welcome to the hell of static dependencies.

Reply:
Thanks, just conquered it! https://github.com/AmigaPorts/devilutionX/pull/42
https://media.tenor.com/images/a5bc6992d129cf564ab2258caa247501/tenor.gif

Reply:
The binary works. But it crashes on loading the character. I assume _BIG_ENDIAN_ wasn't defined. (should be defined by toolchain)

Reply:
And here is the first build from the build server: https://2111-143324737-gh.circle-artifacts.com/0/devilutionx_m68k